ruby/ql/lib/codeql/ruby/ApiGraphs.qll

```diff
@@ -539,27 +579,12 @@ module API {
     /** Gets a node reachable from a use-node. */
     private DataFlow::LocalSourceNode useCandFwd() { result = useCandFwd(TypeTracker::end()) }
 
-    private DataFlow::Node useCandRev(TypeBackTracker tb) {
```

Comment:
After applying the changes in this PR, it didn't perform well anymore.

The pipeline looked like this on `openproject` (manually aborted after a while)
```
(310s) Tuple counts for ApiGraphs#3116a2f3::API::Impl::useNodeStep#3#ffff/4@022f8a1e after 3m48s:
189468     ~0%      {2} r1 = JOIN ApiGraphs#3116a2f3::API::Impl::useCandRev#1#ff_10#join_rhs WITH DataFlowPrivate#462ff392::Cached::isLocalSourceNode#1#f ON FIRST 1 OUTPUT Lhs.1, Lhs.0 'result'
189468     ~0%      {3} r2 = JOIN r1 WITH num#TypeTracker#6f8aa30b::MkTypeBackTracker#fff_2#join_rhs ON FIRST 1 OUTPUT Lhs.1 'result', Lhs.1 'result', Lhs.0
51903      ~3%      {4} r3 = JOIN r2 WITH TypeTracker#6f8aa30b::Cached::stepCall#3#fff_102#join_rhs ON FIRST 1 OUTPUT Lhs.2, Lhs.1 'result', Rhs.1 'mid', Rhs.2
49779      ~2%      {4} r4 = JOIN r2 WITH TypeTracker#6f8aa30b::Cached::stepNoCall#3#fff_102#join_rhs ON FIRST 1 OUTPUT Lhs.2, Lhs.1 'result', Rhs.1 'mid', Rhs.2
101682     ~0%      {4} r5 = r3 UNION r4
189161     ~6%      {4} r6 = JOIN r5 WITH TypeTracker#6f8aa30b::TypeBackTracker::getACompatibleTypeTracker#0#dispred#ff ON FIRST 1 OUTPUT Lhs.3, Rhs.1 't', Lhs.1 'result', Lhs.2 'mid'
1322764717 ~12%     {4} r7 = JOIN r6 WITH TypeTracker#6f8aa30b::Cached::append#2#fff_120#join_rhs ON FIRST 2 OUTPUT Lhs.3 'mid', Rhs.2 'tmid', Lhs.1 't', Lhs.2 'result'
                    return r7
```

When I reverted to idiomatic type tracking, it seemed to perform quite well again, as witnessed by the performance evaluation.

[This evaluation](https://github.com/github/codeql-dca-main/tree/data/asgerf/summarize-loads-trackUseNodeComparison/reports) shows the effect of the commit that reverts to idiomatic type tracking (i.e. the whole branch without that commit, versus the brach with that commit).

We could of course try to hack on the join ordering again, but I couldn't find an example where the pruning seemed beneficial. If you remember a specific project where the pruning was important, please let me know and I'll investigate.

Reply:
FWIW, `useNodeStep` only [got introduced recently](https://github.com/github/codeql/pull/10141), and I was never convinced that it was right to join with `append` first. I wonder what it would look like if that PR got reverted.

Reply:
Reverting my commit and then reverting that PR seems to handle `openproject` without blowing up, albeit with larger tuple counts than with idiomatic type-tracking. I've [started an evaluation](https://github.com/github/codeql-dca-main/issues/7299) of the two branches to see which performs better.

Reply:
Evaluation was messed up, trying again [here](https://github.com/github/codeql-dca-main/issues/7328).

Reply:
[Evaluation](https://github.com/github/codeql-dca-main/issues/7334) comparing pruning with https://github.com/github/codeql/pull/10141 reverted, versus idiomatic type-tracking with no pruning.

The evaluation does not show any benefit from pruning. It's possible that it originally capitalised on incompleteness, and as the call graph and API graph are now more complete, many more nodes can now be tracked to a use site.